### PR TITLE
Throw IOE if service collection is read-only

### LINKED ISSUE #33812: WebApplicationBuilder.Services should fail if accessed after Build()
WebApplicationBuilder.Services does nothing after the application has already been built, so anything trying to modify it is likely buggy. I suppose it possible to want to access this just to read the service descriptors but that feels unlikely.

In addition to throwing InvalidOperationExceptions from the WebApplicationBuilder.Services getter and the various methods on IServiceCollection, we could also consider writing an analyzer similar to the UseAuthorizationAnalyzer to fail at build time if WebApplicationBuilder.Services is obviously incorrectly accessed after build.

### PR DESCRIPTION
**Pull request by captainsafia**
Fixes #33812.

### CODE REVIEW

**Comment by davidfowl** [comment]
Can you make sure to test orchard with this? I'm not convinced we should do this change anymore

**Comment by captainsafia** [comment]
> Can you make sure to test orchard with this? I'm not convinced we should do this change anymore

Can do!

File: src/DefaultBuilder/src/WebApplicationBuilder.cs [hunk 1]
```diff
@@ -138,6 +138,9 @@ public WebApplication Build()
             ((IConfigurationBuilder)Configuration).Sources.Clear();
             Configuration.AddConfiguration(_builtApplication.Configuration);
 
+            // Mark the service collection as read-only to prevent future modifications
+            ((WebApplicationServiceCollection)Services).IsReadOnly = true;
```
Comment (davidfowl) [inline src/DefaultBuilder/src/WebApplicationBuilder.cs:142]: No need to cast, we have a reference to the WebApplicationServiceCollection:

```suggestion
            _services.IsReadOnly = true;
```

File: src/DefaultBuilder/src/WebApplicationServiceCollection.cs [hunk 2]
```diff
@@ -19,17 +19,21 @@ internal sealed class WebApplicationServiceCollection : IServiceCollection
```
Comment (davidfowl) [inline src/DefaultBuilder/src/WebApplicationServiceCollection.cs:19]: The setter is missing the `CheckServicesAccess` call.

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs [hunk 3]
```diff
@@ -669,6 +669,15 @@ public void WebApplicationBuilder_EnablesServiceScopeValidationByDefaultInDevelo
             Assert.ThrowsAny<Exception>(() => builder.Build());
         }
 
+        [Fact]
+        public async Task WebApplicationBuilder_ThrowsExceptionIfServicesAlreadyBuilt()
+        {
+            var builder = WebApplication.CreateBuilder();
+            await using var app = builder.Build();
+
+            Assert.Throws<InvalidOperationException>(() => builder.Services.AddSingleton<IServer>(new MockAddressesServer()));
```
Comment (davidfowl) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:678]: This test needs more asserts (one for each type of mutation).

Comment (captainsafia) [reply src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs]: Added a few more assertions here. I think it gives us enough coverage.

FWIW: All of these map to the `TryAdd` extension method which map to `IServiceCollection.Add` internally so it all converges to the same codepath.

Comment (davidfowl) [reply src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs]: You need a Remove and an indexer write.

**Review by davidfowl — APPROVED** [review]

**Review by davidfowl — APPROVED** [review]

**Pull request merged**